src/materials/vgl-shader-material.js

```diff
@@ -0,0 +1,171 @@
+import { ShaderMaterial } from 'three';
+import { VglMaterialWithMap } from '../mixins';
+import { boolean, number, string } from '../validators';
+
+/**
+ * A material for drawing geometries with custom vertex or fragment shaders, corresponding to
+ * [THREE.ShaderMaterial](https://threejs.org/docs/index.html#api/materials/ShaderMaterial).
+ *
+ * Properties of [VglMaterial](vgl-material) are also available as a mixin.
+ *
+ * Note that some of these properties are not efficient to update. See
+ * [here](https://threejs.org/docs/#manual/en/introduction/How-to-update-things) for some caveats
+ * about updating material state.
+ */
+
+export default {
+  mixins: [VglMaterialWithMap],
+
+  props: {
+    /**
+     * Defines custom constants using `#define` directives within the GLSL code for both the vertex
+     * shader and the fragment shader; each key/value pair yields another directive:
+     * ```
+     * defines: {
+     *     FOO: 15,
+     *     BAR: true
+     * }
+     * ```
+     * yields the lines
+     * ```
+     * #define FOO 15
+     * #define BAR true
+     * ```
+     * in the GLSL code.
+     * @default {}
+     */
+    defines: { type: Object, default: () => ({}) },
+    /**
+     * Define whether the material color is affected by global fog settings; true to pass fog
+     * uniforms to the shader. Note that changing this value will cause the material to be
+     * reconstructed, so be aware of performance if using this reactively. Also note that when using
+     * this, THREE expects fog-related uniforms to be defined on your material; you can use
+     * something like the following to include them:
+     * ```
+     * uniforms: THREE.UniformsUtils.merge([
+     *    THREE.UniformsLib['fog'],
+     *    { other uniforms... }
+     * ]),
+     * ```
+     */
+    fog: { type: boolean, default: false },
+    /**
+     * Fragment shader GLSL code. This is the actual code for the shader.
+     * @default The default fragment shader provided by three.js
+     */
+    fragmentShader: { type: string },
+    /**
+     * Defines whether this material uses lighting; true to pass uniform data related to lighting to
+     * this shader. Note that changing this value will cause the material to be reconstructed, so be
+     * aware of performance if using this reactively. Also note that when using this, THREE expects
+     * lighting-related uniforms to be defined on your material; you can use something like the
+     * following to include them:
+     * ```
+     * uniforms: THREE.UniformsUtils.merge([
+     *    THREE.UniformsLib['lights'],
+     *    { other uniforms... }
+     * ]),
+     * ```
+     * This prop is only used for initialization; it is not reactive.
```

Comment:
@Gollum999 This comment of a line should also be removed.

Reply:
Whoops, good catch.  Removed.

Reply:
Thank you, very fast response!